distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java

```diff
@@ -951,6 +1002,163 @@ public void testCswCreateEventEndpoint() throws Exception {
+        RestitoRetrievalResponse slowFail = new RestitoRetrievalResponse("testData", 500, 3);
+
+        cswServer.whenHttp()
+                .match(Condition.get("/services/csw"),
+                        Condition.parameter("request", "GetRecordById"))
+                .then(getCswRetrievalHeaders(), slowFail.action());
+
+        String restUrl =
+                REST_PATH.getUrl() + "sources/" + CSW_STUB_SOURCE_ID + "/" + CSW_STUB_RESOURCE_ID
+                        + "?transform=resource";
+
+        // Verify that product retrieval fails from the csw stub server.
+        // @formatter:off
+        when().get(restUrl).then().log().all().assertThat().statusCode(500).contentType("text/plain")
+                .body(containsString("cannot retrieve product"));
+        // @formatter:on
+    }
+
+    @Test
+    public void testDownloadFromCacheIfAvailable() throws Exception {
+        RestitoRetrievalResponse restitoRetrievalResponse =
+                new RestitoRetrievalResponse("testData");
+
+        cswServer.whenHttp()
+                .match(Condition.get("/services/csw"),
+                        Condition.parameter("request", "GetRecordById"))
+                .then(getCswRetrievalHeaders(), restitoRetrievalResponse.action());
+
+        String restUrl =
+                REST_PATH.getUrl() + "sources/" + CSW_STUB_SOURCE_ID + "/" + CSW_STUB_RESOURCE_ID
+                        + "?transform=resource";
+
+        // Download product twice, should only call the stub server to download once
+        restitoRetrievalResponse.resetCallCount();
+        // @formatter:off
+        when().get(restUrl).then().log().all().assertThat().contentType("text/plain")
+                .body(is("testData"));
+        when().get(restUrl).then().log().all().assertThat().contentType("text/plain")
+                .body(is("testData"));
+        // @formatter:on
+        assertThat(restitoRetrievalResponse.getCallCount(), equalTo(1));
+    }
+
+    @Test
+    public void testCacheIsUpdatedIfRemoteProductChanges() throws Exception {
+        RestitoRetrievalResponse restitoRetrievalResponse =
+                new RestitoRetrievalResponse("testData");
+
+        cswServer.whenHttp()
+                .match(Condition.get("/services/csw"),
+                        Condition.parameter("request", "GetRecordById"))
+                .then(getCswRetrievalHeaders(), restitoRetrievalResponse.action());
+
+        String restUrl =
+                REST_PATH.getUrl() + "sources/" + CSW_STUB_SOURCE_ID + "/" + CSW_STUB_RESOURCE_ID
+                        + "?transform=resource";
+
+        // Download product twice, and change metacard on stub server between calls.
+        restitoRetrievalResponse.resetCallCount();
+        // @formatter:off
+        when().get(restUrl).then().log().all().assertThat().contentType("text/plain")
+                .body(is("testData"));
+        cswServer.setQueryResponse("csw-query-response-modified.xml");
```

Comment:
Do you think we could use the same response file, have a replacement for the modified time and use `Library.getFileContent()` to replace it?